frontend/src/pages/dashboard/EntitySelector.tsx

```diff
@@ -20,26 +21,123 @@ import {
-import { formatAsDate } from "../../utils/utils";
+import { formatAsDate, isDefined, Maybe } from "../../utils/utils";
 import { Button } from "../../components/Button/Button";
 import { ContextItem } from "../../components/ContextActions/ContextItems";
 
 interface SelectorProps {
   className?: string;
+  experiments?: Experiment[];
 }
 
-export function ProjectSelector({ className }: SelectorProps): React.ReactElement {
+const projectsFilter = (
+  compound: Maybe<CompoundDto>,
+  target: Maybe<TargetDto>,
+  experimentsF: Experiment[]
+): Set<string> =>
+  new Set(
+    experimentsF
+      .filter(
+        experiment =>
+          (isDefined(compound) ? experiment.compoundName === compound.item.name : true) &&
+          (isDefined(target) ? experiment.targetName === target.item.name : true)
+      )
+      .map(experiment => experiment.projectName)
+  );
+
+const compoundsFilter = (
+  project: Maybe<ProjectDto>,
+  target: Maybe<TargetDto>,
+  experimentsF: Experiment[]
+): Set<string> =>
+  new Set(
+    experimentsF
+      .filter(
+        experiment =>
+          (isDefined(project) ? experiment.projectName === project.item.name : true) &&
+          (isDefined(target) ? experiment.targetName === target.item.name : true)
+      )
+      .map(experiment => experiment.compoundName)
+  );
+
+const targetsFilter = (
+  project: Maybe<ProjectDto>,
+  compound: Maybe<CompoundDto>,
+  experimentsF: Experiment[]
+): Set<string> =>
+  new Set(
+    experimentsF
+      .filter(
+        experiment =>
+          (isDefined(project) ? experiment.projectName === project.item.name : true) &&
+          (isDefined(compound) ? experiment.compoundName === compound.item.name : true)
+      )
+      .map(experiment => experiment.targetName)
+  );
+
+export function ProjectSelector({ className, experiments }: SelectorProps): React.ReactElement {
+  const filterNeeds = useRecoilValueLoadable(
+    waitForAll([compoundSelectedQuery, targetSelectedQuery])
+  );
   // TODO make it async one day?
   const projectsL = useRecoilValueLoadable(projectsQuery({}));
   const projectSelectedL = useRecoilValueLoadable(projectSelectedQuery);
   const setProjectSelected = useSetRecoilState(projectSelectedIdAtom);
+  const setCompoundSelected = useSetRecoilState(compoundIdSelectedAtom);
+  const setTargetSelected = useSetRecoilState(targetIdSelectedAtom);
   const setModalDialog = useSetRecoilState(modalDialogAtom);
 
+  const projectsConstruct = useCallback(
+    (projects: ProjectDto[]) => {
+      if (filterNeeds.state === "hasValue" && filterNeeds.contents && experiments) {
+        const [selectedCompound, selectedTarget] = filterNeeds.contents;
+        const projectNames = projectsFilter(selectedCompound, selectedTarget, experiments);
+
+        const [filteredProjects, restProjects] = projects.reduce(
+          (acc: ProjectDto[][], project) => {
+            acc[projectNames.has(project.item.name) ? 0 : 1].push(project);
```

Comment:
honestly, looks like hujita. you could pattern-match on acc like:
`[filteredProjects, restProjects]` and make an if instead of this creepy statement